src/frontend/src/binder/expr/value.rs

```diff
@@ -132,27 +132,92 @@ impl Binder {
         Ok(expr)
     }
 
-    pub(super) fn bind_array_cast(&mut self, exprs: Vec<Expr>, ty: DataType) -> Result<ExprImpl> {
-        let inner_type = if let DataType::List(datatype) = &ty {
-            *datatype.clone()
-        } else {
-            return Err(ErrorCode::BindError(format!(
-                "cannot cast array to non-array type {}",
-                ty
-            ))
-            .into());
-        };
+    pub(super) fn bind_map(&mut self, entries: Vec<(Expr, Expr)>) -> Result<ExprImpl> {
+        if entries.is_empty() {
+            return Err(ErrorCode::BindError("cannot determine type of empty map\nHINT:  Explicitly cast to the desired type, for example MAP{}::map(int,int).".into()).into());
+        }
+        let mut keys = Vec::with_capacity(entries.len());
+        let mut values = Vec::with_capacity(entries.len());
+        for (k, v) in entries {
+            keys.push(self.bind_expr_inner(k)?);
+            values.push(self.bind_expr_inner(v)?);
+        }
+        let key_type = align_types(keys.iter_mut())?;
+        let value_type = align_types(values.iter_mut())?;
+
+        let keys: ExprImpl = FunctionCall::new_unchecked(
+            ExprType::Array,
+            keys,
+            DataType::List(Box::new(key_type.clone())),
+        )
+        .into();
+        let values: ExprImpl = FunctionCall::new_unchecked(
+            ExprType::Array,
+            values,
+            DataType::List(Box::new(value_type.clone())),
+        )
+        .into();
+
+        let expr: ExprImpl = FunctionCall::new_unchecked(
+            ExprType::MapFromEntries,
+            vec![keys, values],
+            DataType::Map(MapType::from_kv(key_type, value_type)),
+        )
+        .into();
+        Ok(expr)
+    }
 
+    pub(super) fn bind_array_cast(
+        &mut self,
+        exprs: Vec<Expr>,
+        element_type: Box<DataType>,
+    ) -> Result<ExprImpl> {
         let exprs = exprs
             .into_iter()
-            .map(|e| self.bind_cast_inner(e, inner_type.clone()))
+            .map(|e| self.bind_cast_inner(e, *element_type.clone()))
             .collect::<Result<Vec<ExprImpl>>>()?;
 
-        let expr: ExprImpl = FunctionCall::new_unchecked(ExprType::Array, exprs, ty).into();
+        let expr: ExprImpl =
+            FunctionCall::new_unchecked(ExprType::Array, exprs, DataType::List(element_type))
+                .into();
         Ok(expr)
     }
 
-    pub(super) fn bind_array_index(&mut self, obj: Expr, index: Expr) -> Result<ExprImpl> {
+    pub(super) fn bind_map_cast(
```

Comment:
It would be good if we can reuse some code with `bind_map`.

Reply:
I considered to use a `Option<MapType>` arg, but I'm not sure that's more beautiful since we will need to add a few branching in the function then.